### PR TITLE
[PUBDEV-4711] Handling of supported and unsupported formats per model

### PR DESCRIPTION
**Pull request by michal-raska**
The idea is, that each model will have a `haveMojo`, `havePojo` flag, which indicates, whether export to particular format is supported on given model. This will propagate also propagate to the clients.

### CODE REVIEW

**Comment by michal-raska** [comment]
CHANGELOG:

* JAVA:
  * `Model`
    * added `have(Pojo | Mojo)` getters, return `false` by default
  * `ModelSchemaBase`
    * added `have_(mojo | pojo)` fields, `false` by default
    * fields are being populated in `fillFromImpl` in `ModelSchema`
  * `Models`
    * overrides of the default `have(Pojo | Mojo)` getters
  * `ModelsHandler`
    * `fetchJavaCode` and `fetchMojo` methods now check if model supports requested format

* PYTHON:
  * `estimator_base`
    * sets the `_have_(pojo | mojo)` fields of model
  * `h2o`
    * check if model supports exporting to POJO before calling backend
  * `model`
    * added `have_(pojo | mojo)` fields and properties 
    * check if model supports exporting to MOJO before calling backend

* TEST

**Comment by michal-raska** [comment]
@michalkurka I need the information in Model. Is it acceptable, if I read it from ModelBuilder and make it accessible in Model?

File: h2o-core/src/main/java/hex/Model.java [hunk 1]
```diff
@@ -1614,7 +1622,7 @@ protected void toJavaPredictBody(SBPrintStream body,
                                    CodeGeneratorPipeline classCtx,
                                    CodeGeneratorPipeline fileCtx,
                                    boolean verboseCode) {
-    throw new IllegalArgumentException("This model type does not support conversion to Java");
+    throw new UnsupportedConversionToJavaException("This model type does not support conversion to Java");
```
Comment (michalkurka) [inline h2o-core/src/main/java/hex/Model.java]: I don't think this should really be an IllegalArgumentException. Better exception seems to be UnsupportedOperationException. If we need a custom exception here, consider inheriting from H2OAbstractRuntimeException.

File: h2o-core/src/main/java/water/api/schemas3/ModelSchemaBaseV3.java [hunk 2]
```diff
@@ -1,5 +1,6 @@
 package water.api.schemas3;
 
+import hex.ModelBuilder;
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/api/schemas3/ModelSchemaBaseV3.java]: Redundant import

File: h2o-core/src/main/java/water/api/RequestServer.java [hunk 3]
```diff
@@ -314,12 +314,18 @@ public void doGeneric(String method, HttpServletRequest request, HttpServletResp
 
       resp.writeTo(response.getOutputStream());
 
+    } catch (Model.UnsupportedConversionToJavaException e) {  // This error type should be passed to clients.
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/api/RequestServer.java]: I still believe there should not be any special handling of just this particular state.

File: h2o-core/src/main/java/hex/Model.java (hunk 1 above)
Comment (michal-raska) [reply h2o-core/src/main/java/hex/Model.java]: throwing `UnsupportedOperationException`

File: h2o-core/src/main/java/water/api/schemas3/ModelSchemaBaseV3.java (hunk 2 above)
Comment (michal-raska) [reply h2o-core/src/main/java/water/api/schemas3/ModelSchemaBaseV3.java]: removed

File: h2o-core/src/main/java/water/api/RequestServer.java (hunk 3 above)
Comment (michal-raska) [reply h2o-core/src/main/java/water/api/RequestServer.java]: Removed entire catch, so only the original with `IOException` remains

File: h2o-core/src/main/java/hex/Model.java [hunk 4]
```diff
@@ -104,6 +104,14 @@ public double defaultThreshold() {
 
   public final boolean isSupervised() { return _output.isSupervised(); }
 
+  public boolean havePojo() {
+    return ModelBuilder.havePojo(_parms.algoName().toLowerCase());
```
Comment (michalkurka) [inline h2o-core/src/main/java/hex/Model.java]: Very minor suggestion: perhaps you can move the toLowerCase to ensureBuilderIndex method?

Comment (michal-raska) [reply h2o-core/src/main/java/hex/Model.java]: 👍

**Comment by michal-raska** [comment]
@wendycwong Thank you very much, I'm very happy you like it :). One more question, should this be merged to `master` only, or `rel-weierstrass` as well? The linked JIRA _doesn't have_ the `Fix Version` tag, so I think it should be merged, to `master` only, just want to make sure.

**Comment by michalkurka** [comment]
@michal-raska this is not a customer reported issue, it doesn't have to go to fix release. IMO you can merge just to master.

**Pull request merged**